src/dtx/dtx_common.c

```diff
@@ -0,0 +1,632 @@
+		       sizeof(struct dtx_id) * commit_cnt);
+		dth->dth_dti_cos_count = dti_cos_count;
+		dth->dth_dti_cos = dti_cos;
+	}
+
+	if (abort_cnt > 0) {
+		rc = dtx_abort(po_uuid, co_uuid, abort_dtes, abort_cnt,
+				version);
+		if (rc == -DER_NONEXIST)
+			rc = 0;
+	}
+
+out:
+	if (commit_ids != NULL)
+		D_FREE(commit_ids);
+	if (abort_dtes != NULL)
+		D_FREE(abort_dtes);
+
+	return rc > 0 ? 0 : rc;
+}
+
+int
+dtx_batched_commit_register(struct ds_cont_hdl *hdl)
+{
+	struct ds_cont			*cont = hdl->sch_cont;
+	struct dtx_batched_commit_args	*dbca;
+	d_list_t			*head;
+
+	D_ASSERT(cont != NULL);
+
+	head = &dss_get_module_info()->dmi_dtx_batched_list;
+	d_list_for_each_entry(dbca, head, dbca_link) {
+		if (uuid_compare(dbca->dbca_cont_hdl->sch_cont->sc_uuid,
+				 cont->sc_uuid) == 0)
+			return 0;
+	}
+
+	cont->sc_closing = 0;
+	cont->sc_deleted = 0;
+
+	D_ALLOC_PTR(dbca);
+	if (dbca == NULL)
+		return -DER_NOMEM;
+
+	ds_cont_hdl_get(hdl);
+	dbca->dbca_cont_hdl = hdl;
+	d_list_add_tail(&dbca->dbca_link, head);
+
+	return 0;
+}
+
+void
+dtx_batched_commit_deregister(struct ds_cont_hdl *hdl)
+{
+	struct ds_cont			*cont = hdl->sch_cont;
+	struct dtx_batched_commit_args	*dbca;
+	d_list_t			*head;
+	ABT_future			 future;
+	int				 rc;
+
+	if (cont == NULL)
+		return;
+
+	if (cont->sc_closing) {
+		if (cont->sc_dtx_flush_cbdata == NULL)
+			return;
+
+		cont->sc_dtx_flush_wait_count++;
+		future = cont->sc_dtx_flush_cbdata;
+		rc = ABT_future_wait(future);
+		D_ASSERTF(rc == ABT_SUCCESS,
+			  "ABT_future_wait failed for DTX flush (1) on "
+			  DF_UUID": rc = %d\n", DP_UUID(cont->sc_uuid), rc);
+
+		if (--(cont->sc_dtx_flush_wait_count) == 0) {
+			cont->sc_dtx_flush_cbdata = NULL;
+			ABT_future_free(&future);
+		}
+
+		return;
+	}
+
+	head = &dss_get_module_info()->dmi_dtx_batched_list;
+	d_list_for_each_entry(dbca, head, dbca_link) {
+		if (uuid_compare(dbca->dbca_cont_hdl->sch_cont->sc_uuid,
+				 cont->sc_uuid) != 0)
+			continue;
+
+		D_ASSERT(hdl == dbca->dbca_cont_hdl);
+
+		/* Notify the dtx_batched_commit ULT to flush the
+		 * committable DTXs via setting @sc_closing as 1.
+		 *
+		 * Then current ULT will wait here until the DTXs
+		 * have been committed by dtx_batched_commit ULT
+		 * that will wakeup current ULT.
+		 */
+		D_ASSERT(cont->sc_dtx_flush_cbdata == NULL);
+		D_ASSERT(cont->sc_dtx_flush_wait_count == 0);
+
+		rc = ABT_future_create(1, NULL, &future);
+		if (rc != ABT_SUCCESS) {
+			D_ERROR("ABT_future_create failed for DTX flush on "
+				DF_UUID" %d\n", DP_UUID(cont->sc_uuid), rc);
+			return;
+		}
+
+		cont->sc_closing = 1;
+
+		cont->sc_dtx_flush_wait_count = 1;
+		cont->sc_dtx_flush_cbdata = future;
+		rc = ABT_future_wait(future);
+		D_ASSERTF(rc == ABT_SUCCESS, "ABT_future_wait failed "
+			  "for DTX flush (2) on "DF_UUID": rc = %d\n",
+			  DP_UUID(cont->sc_uuid), rc);
+
+		if (--(cont->sc_dtx_flush_wait_count) == 0) {
+			cont->sc_dtx_flush_cbdata = NULL;
+			ABT_future_free(&future);
+		}
```

Comment:
the above ABT_future_wait and ABT_future_free are dupped in this function, how about we just break at L557, and move this whole code block out from the loop, and calls goto at L539 so they can share the same ABT_future_wait/free at the end of this function?

Reply:
OK